### PR TITLE
Test Ruby 3.4.0-preview2

### PR DESCRIPTION
**Pull request by kaylareopelle**
New preview version out! 🎉 

[Release](https://github.com/ruby/setup-ruby/releases/tag/v1.196.0)

🟢 CI: https://github.com/newrelic/newrelic-ruby-agent/actions/runs/11263849338

### CODE REVIEW

**Comment by kaylareopelle** [comment]
Two test failures with the new version. 

---

**Unit tests:**
NewRelic::Agent::Configuration::ManagerTest#test_logger_does_not_receive_excluded_settings [/opt/hostedtoolcache/Ruby/3.4.0-preview2/x64/lib/ruby/gems/3.4.0+0/gems/minitest-5.3.3/lib/minitest/assertions.rb:129]:
[82](https://github.com/newrelic/newrelic-ruby-agent/actions/runs/11241665388/job/31256212152#step:12:83)
 rails test /home/runner/work/newrelic-ruby-agent/newrelic-ruby-agent/test/new_relic/agent/configuration/manager_test.rb:533
[372](https://github.com/newrelic/newrelic-ruby-agent/actions/runs/11241665388/job/31256212152#step:12:373)

---

**multiverse config_file_loading suite**
ruby: prism/util/pm_newline_list.c:93: pm_newline_list_line_column: Assertion `cursor >= list->start' failed.
[1349](https://github.com/newrelic/newrelic-ruby-agent/actions/runs/11241665388/job/31256213418#step:12:1350)

Running "config_file_loading" using CHAIN for Envfile entry 1
[1286](https://github.com/newrelic/newrelic-ruby-agent/actions/runs/11241665388/job/31256213418#step:12:1287)
Starting tests in child PID 21523 at 2024-10-08 18:43:19 +0000
[1287](https://github.com/newrelic/newrelic-ruby-agent/actions/runs/11241665388/job/31256213418#step:12:1288)
Bundling in /home/runner/work/newrelic-ruby-agent/newrelic-ruby-agent/test/multiverse/suites/config_file_loading...
[1288](https://github.com/newrelic/newrelic-ruby-agent/actions/runs/11241665388/job/31256213418#step:12:1289)
WARN: Unresolved or ambiguous specs during Gem::Specification.reset:
[1289](https://github.com/newrelic/newrelic-ruby-agent/actions/runs/11241665388/job/31256213418#step:12:1290)
      stringio (>= 0)
[1290](https://github.com/newrelic/newrelic-ruby-agent/actions/runs/11241665388/job/31256213418#step:12:1291)
      Available/installed versions of this gem:
[1291](https://github.com/newrelic/newrelic-ruby-agent/actions/runs/11241665388/job/31256213418#step:12:1292)
      - 3.1.2.dev
[1292](https://github.com/newrelic/newrelic-ruby-agent/actions/runs/11241665388/job/31256213418#step:12:1293)
      - 3.1.1

**Theory:** The switch to the prism parser is causing a problem for the [fakefs gem](https://github.com/fakefs/fakefs/) used in this suite. The gem hasn't been updated since 2014. We may need to add an explicit stringio install in the Envfile.

**Comment by kaylareopelle** [comment]
A test failure still exists on the CI that I'm not able to reproduce locally. It seems like a coincidence that the failure is also in `config_file_loading`. I'll try using a different bundler version tomorrow. 

```
Running "config_file_loading" using CHAIN for Envfile entry 2
Starting tests in child PID 58391 at 2024-10-09 00:04:28 +0000
Bundling in /home/runner/work/newrelic-ruby-agent/newrelic-ruby-agent/test/multiverse/suites/config_file_loading...
WARN: Unresolved or ambiguous specs during Gem::Specification.reset:
      stringio (>= 0)
      Available/installed versions of this gem:
      - 3.1.2.dev
      - 3.1.1
WARN: Clearing out unresolved specs. Try 'gem cleanup <gem>'
Please report a bug if this causes problems.
Fetching gem metadata from https://rubygems.org/........
Resolving dependencies...
Using rake 13.2.1
Using bundler 2.4.1
Using base64 0.2.0
Using bigdecimal 3.1.8
Using fakefs 2.5.0
Using metaclass 0.0.4
Using minitest 5.16.3
Using minitest-stub-const 0.6
Using mutex_m 0.2.0
Using newrelic_rpm 9.14.0 from source at `/home/runner/work/newrelic-ruby-agent/newrelic-ruby-agent` and installing its executables
Using ostruct 0.6.0
Using psych 3.3.4
Using warning 1.4.0
Using webrick 1.8.2
Using mocha 1.9.0
Bundle complete! 13 Gemfile dependencies, 15 gems now installed.
Use `bundle info [gemname]` to see where a bundled gem is installed.
ruby: prism/util/pm_newline_list.c:93: pm_newline_list_line_column: Assertion `cursor >= list->start' failed.
"config_file_loading" for Envfile entry 2 failed!
```

**Comment by kaylareopelle** [comment]
Explicitly installing stringio resolved the following warning: 
> WARN: Unresolved or ambiguous specs during Gem::Specification.reset:
      stringio (>= 0)
      Available/installed versions of this gem:
      - 3.1.2.dev
      - 3.1.1
WARN: Clearing out unresolved specs. Try 'gem cleanup <gem>' 

But, the suite still fails because of what looks like a prism error: 
> ruby: prism/util/pm_newline_list.c:93: pm_newline_list_line_column: Assertion `cursor >= list->start' failed.
"config_file_loading" for Envfile entry 2 failed! 

I'd like to just skip the `config_file_loading` suite until a new preview version is released to see if there's a bug within Prism that, when fixed, will allow the suite to run.

File: test/multiverse/suites/config_file_loading/Envfile [hunk 1]
```diff
@@ -10,14 +10,23 @@ PSYCH_VERSIONS = [
   ['3.3.0', 2.4]
 ]
 
+def stringio_version
```
Comment (kaylareopelle) [inline test/multiverse/suites/config_file_loading/Envfile:13]: The CI still hits errors related to Prism for this suite only. I'd like to skip the suite for Ruby 3.4 until the next preview release to see if the problem is something that might get resolved within the Prism internals.

**Review by fallwith — DISMISSED** [review]

File: test/new_relic/agent/configuration/manager_test.rb [hunk 2]
```diff
@@ -533,8 +533,20 @@ def test_unsatisfied_values_stay_cached
...
+      # Ruby 3.4+:
+      # app_name: \"test\"
+      # Remove the condition once we support only version 3.4+
+      if Gem::Version.new(RUBY_VERSION) >= Gem::Version.new('3.4.0')
+        assert_includes(log, 'app_name:')
+        refute_includes(log, 'license_key:')
+      else
+        assert_includes(log, ':app_name')
+        refute_includes(log, ':license_key')
+      end
```
[the remark this answers is not in the document]
Comment (fallwith) [reply test/new_relic/agent/configuration/manager_test.rb:549]: That's a good idea.

File: test/agent_helper.rb [hunk 3]
```diff
@@ -112,7 +112,7 @@ def assert_log_contains(log, message)
   lines = log.array
 
   assert (lines.any? { |line| line.match(message) }),
-    "Could not find message. Log contained: #{lines.join("\n")}"
+    "Could not find message: '#{message.inspect}'. Log contained: #{lines.join("\n")}"
```
[the remark this answers is not in the document]
Comment (fallwith) [reply test/agent_helper.rb:115]: Keeping it sounds good.

File: lib/new_relic/agent/configuration/default_source.rb [hunk 4]
```diff
@@ -427,6 +427,7 @@ def self.enforce_fallback(allowed_values: nil, fallback: nil)
           :public => true,
           :type => String,
           :allowed_from_server => false,
+          :exclude_from_reported_settings => true,
```
[the remark this answers is not in the document]
Comment (fallwith) [reply lib/new_relic/agent/configuration/default_source.rb:430]: It sounds like a good candidate for excluding, yep.

**Review by fallwith — APPROVED** [review]

File: test/new_relic/agent/configuration/manager_test.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (kaylareopelle) [reply test/new_relic/agent/configuration/manager_test.rb]: Cool! Updated in c0091a31848a6008eac03e9bf0b7fac03d26eee2

**Review by tannalynn — APPROVED** [review]

**Pull request merged**